### ISSUE TITLE
Preprocessor output should use ANSI colors for errors/warning

### ISSUE THREAD

**Issue by edsko** [well-typed/hs-bindgen#84]

**Comment by dschrempf** [comment]
If accepted, this should be covered by #647. Do you want to color the complete log message, or only the "Log Level Tag" (e.g., Warning, or Error)? @edsko

**Comment by edsko** [comment]
Oh, nice :) I don't have a strong preference here, whatever you think is most clear :)

**Comment by TravisCardwell** [comment]
I hope that this only outputs ANSI when writing to a terminal.  Logging that always output ANSI colors even when writing to a file or pipe can be very annoying. 😅

**Comment by dschrempf** [comment]
Yes, the logging can be configured.

By default, `mkTracerStdOut` uses ANSI codes, and `withTracerFile` does not use ANSI codes.

**Comment by dschrempf** [comment]
I just did some research, and when we want to automatically determine within `hs-bindgen` if `stdout` supports ANSI, or not, we need to make `mkTracerStdOut` a `IO ()` action. We could remedy this, by just providing `withTraceXYZ` functions in IO.

**Comment by TravisCardwell** [comment]
Indeed.  I usually use [`ansi-terminal`](https://hackage.haskell.org/package/ansi-terminal-1.1.2/docs/System-Console-ANSI.html#g:35) because it is cross-platform and checks the `TERM` environment variable, which I have used many times to avoid problematic ANSI codes.  I wonder if that package is already in our transitive dependency graph...

**Comment by dschrempf** [comment]
Yes, I added `ansi-terminal`. I think it was not in the transitive dependencies, because only tests use `ansi-diff`. (But maybe some other library was alread using it).

This is how `withTracerStdOut` looks like: https://github.com/well-typed/hs-bindgen/blob/9d9194f5cd0b5a223d150a7aad18700c815c819c/hs-bindgen/src-internal/HsBindgen/Util/Tracer.hs#L114

And it is used like so: https://github.com/well-typed/hs-bindgen/blob/9d9194f5cd0b5a223d150a7aad18700c815c819c/hs-bindgen/app/hs-bindgen-cli.hs#L21

**Comment by dschrempf** [comment]
I think this done and can be closed, but I may be mistaken.

**Comment by edsko** [comment]
Agreed.
